### PR TITLE
change register schema from client to metaserver

### PR DESCRIPTION
**Pull request by songtong**

### CODE REVIEW

File: hermes-core/src/main/java/com/ctrip/hermes/core/meta/internal/LocalMetaProxy.java [hunk 1]
```diff
@@ -62,4 +62,14 @@ public LeaseAcquireResponse tryAcquireBrokerLease(String topic, int partition, S
 		return new ArrayList<SubscriptionView>();
 	}
 
+	@Override
+	public int registerSchema(String schema, String subject) {
+		return 0;
```
Comment (yiming187) [inline hermes-core/src/main/java/com/ctrip/hermes/core/meta/internal/LocalMetaProxy.java]: default -1 is better

File: hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java [hunk 2]
```diff
@@ -264,4 +264,32 @@ public String apply(String ip) {
 		}
 	}
 
+	@Override
+	public int registerSchema(String schema, String subject) {
+		Map<String, String> params = new HashMap<>();
+		params.put("schema", schema);
+		params.put("subject", subject);
+		String response = post("/schema/register", null, params);
```
Comment (yiming187) [inline hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java]: put IO operation(post) in try block.

File: hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java [hunk 3]
```diff
@@ -264,4 +264,32 @@ public String apply(String ip) {
...
+		if (response != null) {
+			try {
+				return Integer.valueOf(response);
+			} catch (Exception e) {
+				log.error("Can not parse response, schema: {}, subject: {}\nResponse: {}", schema, subject, response);
```
Comment (yiming187) [inline hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java]: you may want to show the exception detail, why the parse failed. The exception may come from the server side.

File: hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java (hunk 2 above)
Comment (yiming187) [inline hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java]: maybe we need update the post/get method, currently, we ignore all exceptions.

File: hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java [hunk 4]
```diff
@@ -264,4 +264,32 @@ public String apply(String ip) {
...
+	@Override
+	public String getSchemaString(int schemaId) {
+		Map<String, String> params = new HashMap<>();
+		params.put("id", String.valueOf(schemaId));
+		String response = get("/schema/register", params);
+		if (response != null) {
+			return response;
+		}
+		log.error("No response while getting meta server[getSchemaString]");
+		return null;
```
Comment (yiming187) [inline hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java]: empty is better than null, null may cause NPE

File: hermes-kafka/src/main/java/com/ctrip/hermes/kafka/codec/assist/HermesKafkaAvroDeserializer.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+	public void setSchemaRestService(HermesSchemaRestService restService) {
+		m_schemaRestService = restService;
+	}
+
+	@Override
+	@SuppressWarnings("unchecked")
+	public void configure(Map<String, ?> configs, boolean isKey) {
+		Map<String, String> m = new HashMap<String, String>();
+		m.put("schema.registry.url", "http://127.0.0.1:8081");
```
Comment (yiming187) [inline hermes-kafka/src/main/java/com/ctrip/hermes/kafka/codec/assist/HermesKafkaAvroDeserializer.java]: we'd better remove the test/temporary code.

File: hermes-kafka/src/main/java/com/ctrip/hermes/kafka/codec/assist/HermesSchemaRestService.java [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+import java.io.IOException;
+import java.util.Map;
+
+import org.unidal.lookup.annotation.Inject;
+import org.unidal.lookup.annotation.Named;
+
+import com.ctrip.hermes.core.meta.internal.MetaManager;
+
+@Named(type = HermesSchemaRestService.class)
+public class HermesSchemaRestService extends RestService {
```
Comment (yiming187) [inline hermes-kafka/src/main/java/com/ctrip/hermes/kafka/codec/assist/HermesSchemaRestService.java]: This class the the rest client, not the rest provider. Could we figure out more meaningful name?

File: hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/rest/resource/SchemaRegisterResource.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+			try {
+				RestService restService = new RestService(codec.findProperty("schema.registry.url").getValue());
+				int id = restService.registerSchema(m.get("schema"), m.get("subject"));
+				return Response.status(Status.OK).entity(id).build();
+			} catch (Exception e) {
+				throw new RestException("Register schema failed.", e);
+			}
+		}
+		if (log.isDebugEnabled()) {
+			log.error("Register schema failed. Codec avro not found.");
```
Comment (yiming187) [inline hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/rest/resource/SchemaRegisterResource.java]: log.error should not be under isDebugEnabled()

File: hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/rest/resource/SchemaRegisterResource.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+		Codec codec = m_metaHolder.getMeta().getCodecs().get("avro");
+		if (codec != null) {
+			try {
+				RestService restService = new RestService(codec.findProperty("schema.registry.url").getValue());
+				return Response.status(Status.OK).entity(restService.getId(id).getSchemaString()).build();
+			} catch (Exception e) {
+				throw new RestException("Get schema string failed.", e);
+			}
+		} else if (log.isDebugEnabled()) {
+			log.error("Get schema string failed. Codec avro not found.");
```
Comment (yiming187) [inline hermes-metaserver/src/main/java/com/ctrip/hermes/metaserver/rest/resource/SchemaRegisterResource.java]: same as above

File: hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java (hunk 2 above)
Comment (songtong) [inline hermes-core/src/main/java/com/ctrip/hermes/core/meta/remote/RemoteMetaProxy.java]: post-operation in MetaProxy has already caught Exception, as well as get-operation

File: hermes-kafka/src/main/java/com/ctrip/hermes/kafka/codec/assist/HermesKafkaAvroDeserializer.java (hunk 5 above)
Comment (songtong) [inline hermes-kafka/src/main/java/com/ctrip/hermes/kafka/codec/assist/HermesKafkaAvroDeserializer.java]: m.put("schema.registry.url", "http://127.0.0.1:8081");

**Pull request merged**